block-storage/src/main/scala/coop/rchain/blockstorage/BlockDagFileStorage.scala

```diff
@@ -1034,6 +1082,121 @@ object BlockDagFileStorage {
+                                    )
+                                  case _ =>
+                                    Sync[F].raiseError[List[(DeployId, BlockHash)]](
+                                      BlockHashesByDeployLogIsCorrupted
+                                    )
+                                }
+                     } yield result
+                   case None =>
+                     result.reverse.pure[F]
+                 }
+      } yield result
+    readRec(List.empty)
+  }
+
+  def calculateBlockHashesByDeployCrc[F[_]: Monad](
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[Crc32[F]] = {
+    val crc = Crc32.empty[F]()
+    blockHashesByDeploy.traverse_[F, Unit] {
+      case (deployId, blockHash) =>
+        crc.update(deployId.size.toByteString.concat(deployId).concat(blockHash).toByteArray)
+    } >> crc.pure[F]
+  }
+
+  private def truncateBlockHashesByDeployLog[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F],
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[Unit] = {
+    val lastRecordSize: Long = BlockHash.Length.toLong + BlockHash.Length.toLong
+    for {
+      length <- randomAccessIO.length
+      _      <- randomAccessIO.setLength(length - lastRecordSize)
+    } yield ()
+  }
+
+  private def validateBlockHashesByDeploy[F[_]: Sync](
+      randomAccessIO: RandomAccessIO[F],
+      readBlockHashesByDeployCrc: Long,
+      blockHashesByDeploy: List[(DeployId, BlockHash)]
+  ): F[(List[(DeployId, BlockHash)], Crc32[F])] =
+    for {
+      fullCalculatedCrc <- calculateBlockHashesByDeployCrc[F](blockHashesByDeploy)
+      result <- Monad[F].ifM(fullCalculatedCrc.value.map(_ == readBlockHashesByDeployCrc))(
+                 (blockHashesByDeploy, fullCalculatedCrc).pure[F],
+                 blockHashesByDeploy match {
+                   case Nil =>
+                     Sync[F].raiseError[(List[(DeployId, BlockHash)], Crc32[F])](
+                       BlockHashesByDeployLogIsCorrupted
+                     )
+                   case _ :: _ =>
+                     // Trying to delete the last log entry
+                     for {
+                       withoutLastCalculatedCrc <- calculateBlockHashesByDeployCrc[F](
+                                                    blockHashesByDeploy.init
+                                                  )
+                       result <- Monad[F].ifM(
+                                  withoutLastCalculatedCrc.value
+                                    .map(_ == readBlockHashesByDeployCrc)
+                                )(
+                                  for {
+                                    _ <- truncateBlockHashesByDeployLog(
+                                          randomAccessIO,
+                                          blockHashesByDeploy
+                                        )
+                                  } yield (blockHashesByDeploy.init, withoutLastCalculatedCrc),
```

Comment:
How big is the list of blockHashesByDeploy? The init method is linear and you are calling it twice so it might be worth to cache the result

Reply:
Good point